sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -242,6 +247,10 @@ public FieldType type() {
     public abstract TypeName getTypeName();
     // For container types (e.g. ARRAY), returns the type of the contained element.
     @Nullable public abstract FieldType getComponentType();
+    // For MAP type, returns the type of the key element.
+    @Nullable public abstract FieldType getComponentKeyType();
+    // For MAP type, returns the type of the value element.
+    @Nullable public abstract FieldType getComponentValueType();
```

Comment:
<!--thread_id:cc_180487493_t; commit:7bf9d2048e8d568ee9280629db391775cc490358; resolved:0-->
<!--section:context-quote-->
> **XuMingmin** wrote:
> I would prefer to have different fields for MAP and ARRAY, if it doesn't cause significant performance issue.

<!--section:body-->
two questions/comments

1. I think the key type be a TypeName instead of a FieldType? Making it a FieldType makes it legal for the key to be a complex value - e.g. they key could be an array type - which doesn't sound very meaningful to me.

2. Have you considered introducing a new key-value type here (pair of TypeName, FieldType)?

Reply:
+1, will change to key as primitive, and value can be primitive/array/map/row